lib/rpc.ts

```diff
@@ -110,39 +111,53 @@ export class Rpc extends EventEmitter implements IForwarderDest {
   constructor(options: {logger?: IRpcLogger, sendMessage?: SendMessageCB,
                         callWrapper?: ICallWrapper} = {}) {
     super();
-    const {logger = console, sendMessage = inactiveSend, callWrapper = plainCall} = options;
+    const {logger = console, sendMessage = null, callWrapper = plainCall} = options;
     this._logger = logger;
-    this._sendMessage = sendMessage;
+    this._sendMessageCB = sendMessage;
     this._callWrapper = callWrapper;
-    this._inactiveQueue = (this._sendMessage === inactiveSend) ? [] : null;
   }
 
   /**
    * To use Rpc, call this for every message received from the other side of the channel.
    */
   public receiveMessage(msg: any): void {
-    if (this._inactiveQueue) {
-      this._inactiveQueue.push(msg);
+    if (!this._sendMessageCB) {
+      this._inactiveRecvQueue.push(msg);
     } else {
       this._dispatch(msg);
     }
   }
 
   /**
-   * Until start() is called, received messages are queued. This gives you an opportunity to
-   * register implementations and add "message" listeners without the risk of missing messages,
+   * Until start() is called, received and sent messages are queued. This gives you an opportunity
+   * to register implementations and add "message" listeners without the risk of missing messages,
    * even if receiveMessage() has already started being called.
    */
   public start(sendMessage: SendMessageCB) {
-    this._sendMessage = sendMessage;
-    if (this._inactiveQueue) {
-      for (const msg of this._inactiveQueue) {
-        this._dispatch(msg);    // We need to be careful not to throw from here.
+    // Message sent by `_dispatch(...)` are appended to the send queue
+    processQueue(this._inactiveRecvQueue, this._dispatch.bind(this));
+    processQueue(this._inactiveSendQueue, sendMessage);
+    this._sendMessageCB = sendMessage;
+    function processQueue(queue: IMessage[], processFunc: (msg: IMessage) => any) {
```

Comment:
Please make it a top-level function. Also type of processFunc can return void (that says that the return value is not used, e.g. typescript will protect against relying on the return value in the body of processQueue)

Reply:
ok